src/bin/pgcopydb/ld_transform.c

```diff
@@ -2609,3 +2669,297 @@ LogicalMessageValueEq(LogicalMessageValue *a, LogicalMessageValue *b)
+ */
+static bool
+fetchGeneratedColumnsForTable(GeneratedColumnsCache *cache,
+							  const char *nspname,
+							  const char *relname,
+							  GeneratedColumnSet **columns)
+{
+	/*
+	 * NULL cache means that we don't have any generated columns in the
+	 * catalog.
+	 */
+	if (cache == NULL)
+	{
+		return false;
+	}
+
+	GeneratedColumnsCache *item = NULL;
+
+	GeneratedColumnsCache_Lookup key = { 0 };
+
+	NORMALIZED_PG_NAMEDATA_COPY(key.nspname, nspname);
+	NORMALIZED_PG_NAMEDATA_COPY(key.relname, relname);
+
+	HASH_FIND(hh, cache, &key, sizeof(GeneratedColumnsCache_Lookup), item);
+
+	if (item == NULL)
+	{
+		return false;
+	}
+
+	if (item->columns == NULL)
+	{
+		log_error("BUG: Table \"%s.%s\" is in the cache but columns are NULL",
+				  nspname, relname);
+		return false;
+	}
+
+	*columns = item->columns;
+
+	log_trace("Table \"%s.%s\" has generated columns", nspname, relname);
+
+	return true;
+}
+
+
+/*
+ * isGeneratedColumn checks whether the given "attname" is a generated column.
+ *
+ * Returns true if the column is generated, false otherwise.
+ *
+ * NOTE: There is no error condition, if the cache is NULL, it means that we
+ * don't have any generated columns in the catalog.
+ */
+static bool
+isGeneratedColumn(GeneratedColumnSet *columns, const char *attname)
+{
+	char attnameNormalized[PG_NAMEDATALEN] = { 0 };
+
+	NORMALIZED_PG_NAMEDATA_COPY(attnameNormalized, attname);
+
+	GeneratedColumnSet *generatedColumns = NULL;
+
+	HASH_FIND_STR(columns, attnameNormalized, generatedColumns);
+
+	if (generatedColumns != NULL)
+	{
+		log_trace("Column \"%s\" is generated", attnameNormalized);
+
+		return true;
+	}
+
+	return false;
+}
+
+
+/*
+ * prepareGeneratedColumnsCache_hook is a callback function that populates the
+ * generated columns cache from the catalog.
+ */
+static bool
+prepareGeneratedColumnsCache_hook(void *ctx, SourceTable *table)
+{
+	StreamSpecs *specs = (StreamSpecs *) ctx;
+	StreamContext *privateContext = &(specs->private);
+	DatabaseCatalog *sourceDB = specs->sourceDB;
+
+	if (!catalog_s_table_fetch_attrs(sourceDB, table))
+	{
+		log_error("Failed to fetch attributes for table \"%s\".%s",
+				  table->nspname, table->relname);
+		return false;
+	}
+
+	GeneratedColumnsCache *item = (GeneratedColumnsCache *)
+								  calloc(1, sizeof(GeneratedColumnsCache));
+
+	if (item == NULL)
+	{
+		log_error(ALLOCATION_FAILED_ERROR);
+		return false;
+	}
+
+	for (int i = 0; i < table->attributes.count; i++)
+	{
+		SourceTableAttribute *attr = &(table->attributes.array[i]);
+
+		if (attr->attisgenerated)
+		{
+			/* Add a generated column to the GeneratedColumnSet */
+			GeneratedColumnSet *generatedColumn = (GeneratedColumnSet *)
+												  calloc(1, sizeof(GeneratedColumnSet));
+			if (generatedColumn == NULL)
+			{
+				log_error(ALLOCATION_FAILED_ERROR);
+				return false;
+			}
+
+			NORMALIZED_PG_NAMEDATA_COPY(generatedColumn->attname, attr->attname);
+			HASH_ADD_STR(item->columns, attname, generatedColumn);
```

Comment:
I think I see now. I am used to building a hash-table over an array or a linked-list, so that I would have a sequential path and a direct lookup path. Here you're saying we only need the hash-table lookup path to access the data, so that we don't even need to maintain a linked-list or anything.

Reply:
That is correct, we have a higher level hash-table which contains tables which has generated columns and with in that, a set of columns which are generated(again a hash-table).